### PR TITLE
Merge branch Feature/aggregate statistics generator into develop

### LINKED ISSUE #21: Aggregate Data Display
As a user, I can view aggregated data based on common attributes.

Acceptance Criteria:
- Dynamic page showing averages and statistics.
- Aggregation based on name, culture, or material.

### PR DESCRIPTION
**Pull request by AdamStanger**
Files added:

- frontend/src/components/StatisticsPage.jsx: Page containing the tabe for all the projectile points in the catalogue
- frontend/src/components/StatisticsModal.jsx: Modal for displaying charts for the generated statistics
- frontend/tests/statisticsPage.unit.test.jsx

Modifications:

- Functions in backend/controllers/aggregateStatisticsGenerator.js changed to be async to fix issue when passing data to the frontend
- Unit tests for those functions updated based on refactor
- Path to statistics page added (localhost:8080/stats)

### CODE REVIEW

File: backend/config/authenticate.js
Comment (lecegues) [inline backend/config/authenticate.js:1]: This file should be deleted (remnants from the merge from develop). This file was moved to the `middleware` folder

File: backend/package.json
Comment (lecegues) [inline backend/package.json:1]: This should **not** get rid of these scripts and dependencies. Could be a mistake from the merge from develop.

File: frontend/package.json
Comment (lecegues) [inline frontend/package.json:1]: @vitest/coverage-v8 package should **not** be touched. Again, probably something missed due to the merge.

File: frontend/src/components/Home.jsx
Comment (lecegues) [inline frontend/src/components/Home.jsx:1]: Was box supposed to be first or not? If I remember, I know that the order of dependencies can sometimes cause MUI to crash... Please verify

File: frontend/tests/setup.js
Comment (lecegues) [inline frontend/tests/setup.js:1]: As long as this addition to `setup.js` does not interfere with other tests, then it is ok. Just noting this because **every test** will be referencing this `setup.js` file.

**Review by lecegues — CHANGES_REQUESTED** [review]
Please view my comments for some changes. 
Additionally, am I correct in assuming that all of the `routes` were changed because they now use the corresponding `helperFile`? Or were the changes to the `route` files accidental from the merge?

**Comment by brj496** [comment]
> Please view my comments for some changes. Additionally, am I correct in assuming that all of the `routes` were changed because they now use the corresponding `helperFile`? Or were the changes to the `route` files accidental from the merge?

each route was changed to use its corresponding helperFile. They are correct as they are right now

File: frontend/src/components/Home.jsx
Comment (brj496) [reply frontend/src/components/Home.jsx:1]: I'm not sure, I don't remember messing with this file, so it may have been changed due to the merge

File: backend/config/authenticate.js
Comment (brj496) [reply backend/config/authenticate.js:1]: Removed authenticate.js from backend config

File: backend/package.json
Comment (brj496) [reply backend/package.json:1]: Made sure to add back removed dependencies

File: frontend/src/components/Home.jsx
Comment (brj496) [reply frontend/src/components/Home.jsx:1]: turns out they are in the right order

File: frontend/tests/setup.js
Comment (brj496) [reply frontend/tests/setup.js:1]: I think Dani is the one who made this file, but all the tests function on my machine.

File: testing/smoke-test/smoke.spec.js
Comment (lecegues) [inline testing/smoke-test/smoke.spec.js:1]: You've accidentally added back the smoke tests that were removed, and accidentally removed the new smoke test(edge cases). Please change this

**Review by lecegues — CHANGES_REQUESTED** [review]

File: testing/smoke-test/smoke.spec.js
Comment (brj496) [reply testing/smoke-test/smoke.spec.js:1]: added back new tests and removed old tests

**Review by lecegues — APPROVED** [review]
Reviewed changes were fixed.

File: backend/routes/artifacts.js [hunk 1]
```diff
@@ -78,19 +30,11 @@ router.post("/", async (req, res) => {
...
-			relations: ["site", "artifactType"],
-		});
-
-		assert.ok(artifacts.length > 0, "No artifacts found.");
-		res.json(artifacts);
-		logger.info("Fetched all artifacts.");
-	} catch (error) {
-		logger.error("Error fetching Artifacts: " + error.message);
-		res.status(500).json({ error: error.message });
+	const response = await artifactsHelper.newArtifact(req);
```
Comment (lecegues) [inline backend/routes/artifacts.js]: GET request, but calling a POST helper

File: backend/routes/artifacts.js [hunk 2]
```diff
@@ -103,20 +47,11 @@ router.get("/", async (req, res) => {
...
-			relations: ["site", "artifactType"],
-		});
-
-		assert.ok(artifact, "Artifact not found");
-		res.json(artifact);
-		logger.info(`Fetched artifact with ID: ${id}`);
-	} catch (error) {
-		logger.error(`Error fetching Artifact with ID ${id}: ${error.message}`);
-		res.status(500).json({ error: error.message });
+	const response = await artifactsHelper.newArtifact(req);
```
Comment (lecegues) [inline backend/routes/artifacts.js]: GET request, calling POST helper

File: backend/routes/artifacts.js [hunk 3]
```diff
@@ -129,48 +64,11 @@ router.get("/:id", async (req, res) => {
...
-		artifact.artifactType = artifactType || artifact.artifactType;
-		artifact.subtype = subtype || artifact.subtype;
-
-		await artifactRepository.save(artifact);
-		res.json(artifact);
-		logger.info(`Updated artifact with ID: ${id}`);
-	} catch (error) {
-		logger.error(`Error updating Artifact with ID ${id}: ${error.message}`);
-		res.status(500).json({ error: error.message });
+	const response = await artifactsHelper.newArtifact(req);
```
Comment (lecegues) [inline backend/routes/artifacts.js]: PUT req, calling a POST helper

File: backend/routes/artifacts.js [hunk 4]
```diff
@@ -183,18 +81,11 @@ router.put("/:id", async (req, res) => {
...
-		const artifactRepository = await myDatabase.getRepository(Artifact);
-		const deleteResult = await artifactRepository.delete(id);
-
-		assert.strictEqual(deleteResult.affected > 0, true, "Artifact not found");
-		res.status(204).send(); // No Content
-		logger.info(`Deleted artifact with ID: ${id}`);
-	} catch (error) {
-		logger.error(`Error deleting Artifact with ID ${id}: ${error.message}`);
-		res.status(500).json({ error: error.message });
+	const response = await artifactsHelper.newArtifact(req);
```
Comment (lecegues) [inline backend/routes/artifacts.js]: DELETE req, using a POST helper

File: backend/routes/crossSections.js [hunk 5]
```diff
@@ -124,18 +92,11 @@ router.put("/:id", async (req, res) => {
...
-		assert(!isNaN(id), "Invalid ID provided for deletion");
-		const crossSectionRepository = await myDatabase.getRepository(CrossSection);
-		const deleteResult = await crossSectionRepository.delete(id);
-		assert(deleteResult.affected > 0, "CrossSection not found for deletion");
-		res.status(204).send(); // No Content
-		logger.info(`Deleted CrossSection with ID: ${id}`);
-	} catch (error) {
-		logger.error(`Error deleting CrossSection with ID ${id}: ${error.message}`);
-		res.status(500).json({ error: error.message });
+	const response = await crossSectionsHelper.updateCrossSection(req);
```
Comment (lecegues) [inline backend/routes/crossSections.js]: DELETE req, using UPDATE helper

File: backend/routes/materials.js [hunk 6]
```diff
@@ -72,22 +50,14 @@ router.get("/", async (req, res) => {
...
-		if (!material) {
-			logger.warn(`Material with ID${id} was not found in the database`);
-			res.json({ message: "Material not found" });
-		} else {
-			res.json(material);
-		}
-	} catch (error) {
-		logger.error("Error fetching Material:", error);
-		res.json({ error: error.message });
+	const response = await materialsHelper.newMaterial(req);
```
Comment (lecegues) [inline backend/routes/materials.js]: GET req, calling a POST helper

File: backend/routes/materials.js [hunk 7]
```diff
@@ -103,35 +73,17 @@ router.get("/:id", async (req, res) => {
...
-
-		material.name = name;
-		material.description = description;
-
-		await myDatabase.getRepository(Material).save(material);
-		res.json(material);
-	} catch (error) {
-		logger.error("Error updating Material:", error);
-		res.json({ error: error.message });
+	const response = await materialsHelper.newMaterial(req);
```
Comment (lecegues) [inline backend/routes/materials.js]: UPDATE req, calling a POST helper

File: backend/routes/materials.js [hunk 8]
```diff
@@ -145,19 +97,14 @@ router.put("/:id", async (req, res) => {
...
-		if (deleteResult.affected > 0) {
-			res.send();
-		} else {
-			logger.warn(`Material with ID${id} was not found in the database`);
-			res.json({ message: "Material not found" });
-		}
-	} catch (error) {
-		logger.error("Error deleting Material:", error);
-		res.json({ error: error.message });
+	const response = await materialsHelper.newMaterial(req);
```
Comment (lecegues) [inline backend/routes/materials.js]: DELETE req, calling a POST helper

File: backend/routes/periods.js [hunk 9]
```diff
@@ -116,20 +79,14 @@ router.put("/:id", async (req, res) => {
...
-		if (deleteResult.affected > 0) {
-			res.send();
-		} else {
-			logger.warn(`Period with ID${id} was not found in the database`);
-			res.json({ message: "Period not found" });
-		}
-	} catch (error) {
-		logger.error("Error deleting Period:", error);
-		res.json({ error: error.message });
+	const response = await periodsHelper.updatePeriod(req);
```
Comment (lecegues) [inline backend/routes/periods.js]: DELETE req, calling UPDATE helper

File: backend/routes/regions.js [hunk 10]
```diff
@@ -34,16 +21,11 @@ router.get("/", async (req, res) => {
...
-	const { name, description } = req.body;
-	try {
-		const regionRepository = await myDatabase.getRepository(Region);
-		const newRegion = regionRepository.create({ name, description });
-		await regionRepository.save(newRegion);
-		res.json(newRegion);
-	} catch (error) {
-		logger.error("Error creating new Region:", error);
-		res.json({ error: error.message });
+	const response = await regionsHelper.getAllRegions();
```
Comment (lecegues) [inline backend/routes/regions.js]: POST req, calling a GET helper

File: backend/routes/regions.js [hunk 11]
```diff
@@ -56,25 +38,14 @@ router.post("/", async (req, res) => {
...
-		} else {
-			logger.warn(
-				`Region to fetch was not found, ID: ${parseInt(req.params.id)}`,
-			);
-			res.send("Region not found");
-		}
-	} catch (error) {
-		logger.error("Error fetching Region:", error);
-		res.json({ error: error.message });
+	const response = await regionsHelper.getAllRegions();
```
Comment (lecegues) [inline backend/routes/regions.js]: SINGLE GET req, calling a GET ALL helper

File: backend/routes/regions.js [hunk 12]
```diff
@@ -87,25 +58,14 @@ router.get("/:id", async (req, res) => {
...
-			RegionToUpdate.description = description;
-			await regionRepository.save(RegionToUpdate);
-			res.json(RegionToUpdate);
-		} else {
-			res.json({ message: "Region not found" });
-		}
-	} catch (error) {
-		logger.error("Error updating Region:", error);
-		res.json({ error: error.message });
+	const response = await regionsHelper.getAllRegions();
```
Comment (lecegues) [inline backend/routes/regions.js]: UPDATE req, calling a GET helper

File: backend/routes/regions.js [hunk 13]
```diff
@@ -118,20 +78,14 @@ router.put("/:id", async (req, res) => {
...
-		if (deleteResult.affected > 0) {
-			res.send();
-		} else {
-			logger.warn(`Region to delete was not found, ID: ${id}`);
-			res.json({ message: "Region not found" });
-		}
-	} catch (error) {
-		logger.error("Error deleting Region:", error);
-		res.json({ error: error.message });
+	const response = await regionsHelper.getAllRegions();
```
Comment (lecegues) [inline backend/routes/regions.js]: DELETE req, calling a GET helper

**Review by lecegues — CHANGES_REQUESTED** [review]
There are some API requests that call different helpers than what is intended. I have outlined which ones need ot be changed

File: backend/routes/artifacts.js (hunk 1 above)
Comment (brj496) [reply backend/routes/artifacts.js]: Fixed Issue, replaced with correct function call

File: backend/routes/artifacts.js (hunk 2 above)
Comment (brj496) [reply backend/routes/artifacts.js]: Fixed Issue, replaced with correct function call

File: backend/routes/artifacts.js (hunk 3 above)
Comment (brj496) [reply backend/routes/artifacts.js]: Fixed Issue, replaced with correct function call

File: backend/routes/artifacts.js (hunk 4 above)
Comment (brj496) [reply backend/routes/artifacts.js]: Fixed Issue, replaced with correct function call

File: backend/routes/crossSections.js (hunk 5 above)
Comment (brj496) [reply backend/routes/crossSections.js]: Fixed Issue, replaced with correct function call

File: backend/routes/materials.js (hunk 6 above)
Comment (brj496) [reply backend/routes/materials.js]: Fixed Issue, replaced with correct function call

File: backend/routes/materials.js (hunk 7 above)
Comment (brj496) [reply backend/routes/materials.js]: Fixed Issue, replaced with correct function call

File: backend/routes/materials.js (hunk 8 above)
Comment (brj496) [reply backend/routes/materials.js]: Fixed Issue, replaced with correct function call

File: backend/routes/periods.js (hunk 9 above)
Comment (brj496) [reply backend/routes/periods.js]: Fixed Issue, replaced with correct function call

File: backend/routes/regions.js (hunk 10 above)
Comment (brj496) [reply backend/routes/regions.js]: Fixed Issue, replaced with correct function call

File: backend/routes/regions.js (hunk 12 above)
Comment (brj496) [reply backend/routes/regions.js]: Fixed Issue, replaced with correct function call

File: backend/routes/regions.js (hunk 11 above)
Comment (brj496) [reply backend/routes/regions.js]: Fixed Issue, replaced with correct function call

File: backend/routes/regions.js (hunk 13 above)
Comment (brj496) [reply backend/routes/regions.js]: Fixed Issue, replaced with correct function call

**Review by lecegues — APPROVED** [review]
Reviewed changes were fixed. Additionally, some tests stopped working which revealed some bugs (#115) and were fixed. Also removed failing `Login.integ.test.jsx`; refer to issue #116. Branch should be good for merge.

**Review by funmise — APPROVED** [review]
Helper files are a good change to the routes
